src/python/pants/engine/environment.py

```diff
@@ -64,10 +67,45 @@ def check_and_set(name: str, value: Optional[str]):
         return FrozenDict(env_var_subset)
 
 
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class EnvironmentRequest:
+    """Requests a subset of the variables set in the environment.
+
+    Requesting only the relevant subset of the environment reduces invalidation caused by unrelated
+    changes.
+    """
+
+    requested: FrozenOrderedSet[str]
+    allowed: Optional[FrozenOrderedSet[str]]
+
+    def __init__(self, requested: Sequence[str], allowed: Optional[Sequence[str]] = None):
+        self.requested = FrozenOrderedSet(requested)
+        self.allowed = None if allowed is None else FrozenOrderedSet(allowed)
+
+
+class Environment(FrozenDict[str, str]):
+    """A subset of the variables set in the environment."""
+
+
 @rule
 async def complete_environment(session_values: SessionValues) -> CompleteEnvironment:
     return session_values[CompleteEnvironment]
 
 
+@rule
+async def environment_subset(
```

Comment:
This can be a non async function which begs the question of why not just a utility. Are you making room here?

Reply:
Good point: can drop the `async`. But the reason to expose this as a `@rule` is that we want to discourage direct access to the `CompleteEnvironment` and expose an API that encourages requesting a filtered copy, because it should allow `@rules` to be cleaned with early-cutoff in more cases. When a `@rule` requests the entire environment, it will be invalidated for any environment change.

Reply:
Could you s/encourage/force/ by just changing the relevant Process field?

Reply:
Possibly...? I think that a majority of users of `Process` right now are constructing an environment for the process by hand, and only some cases are pulling it in from the outside world. Most users of `InteractiveProcess` will use the `CompleteEnvironment`, but not all, etc.